src/rpc_connection.c

```diff
@@ -772,8 +772,14 @@ rpc_close(rpc_connection_t conn)
 	conn->rco_aborted = true;
 	conn->rco_closed = true;
 	
-	if (conn->rco_error == NULL)
+	if (conn->rco_error == NULL) {
 		conn->rco_error = rpc_get_last_error();
+		/* if not NULL, cleanup can't release this */
+		if (conn->rco_error != NULL)
+			conn->rco_error  = rpc_error_create(
+			    rpc_error_get_code(conn->rco_error),
+			    rpc_error_get_message(conn->rco_error), NULL);
+	}
```

Comment:
What happens when there is no error? rpc_get_last_error() always reports _last_ error. Successful operations never clear it out so it is unclar by looking at this diff whether it's possible to get a stale, unrelated error object. Also, did you mean rpc_copy() here by recreating exactly the same error object?

Reply:
In reverse order - rpc_copy is a good suggestion, I will change it.
In the transport area, a failure generally ends the connection. But the failure may be in another thread, and a failure in a transport thread generally ends up in rpc_close(). If there is no error, there is no problem. If the error was already set into the connection structure, there is no problem, that will be used. If something fails and closes the connection by setting an rpc error, and it is a client-side connection, that will be conveyed pack to the client app (the caller of rpc_client_close()) through the connection structure.  The client may/may not choose to look at it. I think this is generally the correct behavior. If you think we often get errors that are left as stale, there may be other places in the code that get stale errors, not a good thing independent of the connection scenario. Or we could set null objects on success, which may add overhead we don't want right now.

Reply:
I don't think there are any other places that unconditionally read stale errors (to my knowledge). rpc_get_last_error() works like errno: it never gets reset on success, so consumers need an external information that an error has occured to know that they have to read it. Here that information appears to be missing. Or to put it in the other words: you have to know upfront whether a current error is stored before calling rpc_get_last_error(). `rpc_get_last_error() != NULL` check makes me doubt that is the case here.

Reply:
The missing piece may simply be that rpc_client_close() needs to return an int rather than a void, 0 or -1, since rpc_client.c already fetches the rpc_error from the connection if it is present. However, that can be a separate fix. This fix is to address the crash that can occur when rpc_release(conn->rco_error) is called.

Reply:
Looking at the transports, it seems likely that any errors set by rpc_set_last_error() would occur during the connection phase. Transport errors post-connect are set directly into the connection. The code to rpc_get_last_error can therefore be ripped out. That leaves the issue of returning transport errors to the client, but that is another PR.